distributed/tests/test_worker.py

```diff
@@ -2944,3 +2944,54 @@ async def test_who_has_consistent_remove_replica(c, s, *workers):
     assert ("missing-dep", f1.key) in a.story(f1.key)
     assert a.tasks[f1.key].suspicious_count == 0
     assert s.tasks[f1.key].suspicious == 0
+
+
+@gen_cluster(client=True)
+async def test_missing_released_zombie_tasks(c, s, a, b):
+    """
+    Ensure that no fetch/flight tasks are left in the task dict of a
+    worker after everything was released
+    """
+    a.total_in_connections = 0
+    f1 = c.submit(inc, 1, key="f1", workers=[a.address])
+    f2 = c.submit(inc, f1, key="f2", workers=[b.address])
+    key = f1.key
+
+    while key not in b.tasks or b.tasks[key].state != "fetch":
+        await asyncio.sleep(0.01)
+
+    await a.close(report=False)
+
+    del f1, f2
+
+    while b.tasks:
+        await asyncio.sleep(0.01)
+
+
+@gen_cluster(client=True)
+async def test_missing_released_zombie_tasks_2(c, s, a, b):
+    a.total_in_connections = 0
+    f1 = c.submit(inc, 1, key="f1", workers=[a.address])
+    f2 = c.submit(inc, f1, key="f2", workers=[b.address])
+
+    while f1.key not in b.tasks:
+        await asyncio.sleep(0)
```

Comment:
```suggestion
        await asyncio.sleep(0.01)
```

Reply:
The zero sleep is actually deliberate to avoid any timing related problems. fetch->flight->memory can be very fast and I don't want to miss the transition. My sleeping `0` we're checking this upon every loop cycle. This is a bit excessive but guarantees us not to miss this

Reply:
Doesn't ``a.total_in_connections = 0`` causes the state on b to be stuck in "fetch" forever?

Reply:
No, the tasks will transition fo `flight` and trigger a `get_data` to worker A. Based on the `total_in_connections` it will then limit the number of concurrent outgoing connections (`outgoing_current_count`) and will reply with a `{"status": "busy"}` message.